File: solutions/hash_table/1-1.py
"""
Problem Number 1.1
Problem : Implement an algorithm to determine if a string has all unique characters.
"""
import array
class hash_table_class(object):

	def __init__ (self):
		self.hash_table = [None]*128

	def calculate_hash(self,char_to_hash):
		return ord(char_to_hash)

	def insert_to_hash (self,string):

		for char_from_string in string:
			hash_value=self.calculate_hash(char_from_string)

			if self.hash_table[hash_value] != None:
				print("There are duplicate characters")
				return
			
			self.hash_table[hash_value]=char_from_string
		
		print("All character are unique")

print("Please enter the ASCII string")
string=input()

if len(string) > 128:
	print("There are duplicate characters")

HT = hash_table_class()
HT.insert_to_hash(string)

File: solutions/hash_table/1-1.cpp
/*
Problem Number 1.1
Problem : Implement an algorithm to determine if a string has all unique characters.
*/

#include <iostream>
#include <string>
using namespace std;
class MyHash
{
    char hashTable[128];

public:
    MyHash()
    {
        hashTable[128] = {};
    }

private:
    int getHashValue(char c)
    {
        return int(c);
    }

public:
    void insertToHash(string s)
    {
        int hashValue;
        for (int i = 0; i < s.size(); i++)
        {
            hashValue = getHashValue(s[i]);
            if (hashTable[hashValue] != NULL)
            {
                std::cout << "\nThere are duplicate characters";
                return;
            }

            hashTable[hashValue] = s[i];
        }
        std::cout << "All character(s) are unique";
    }
};

int main()
{
    string s;
    std::cout << "\nPlease enter the ASCII string : ";
    std::cin >> s;

    //Total characters in ASCII is 128. So, If a string contains more than 128 chars then there are duplicates
    if (s.size() > 128)
    {
        std::cout << "There are duplicate characters";
        return 1;
    }

    MyHash MH;

    MH.insertToHash(s);

}
